<font color='DarkBlue' size=+2><b>Комментарий ревьюера</b></font><br>
<font color='DarkBlue'>Привет, Андрей! <br>Меня зовут Олег Мазуренко, и я буду проверять этот проект. Спасибо за проделанную работу! <br>Ко мне можно обращаться на «ты». Как мне обращаться, напиши.</font><br>
<font color='DarkBlue'>Я буду использовать различные цвета, чтобы было удобнее воспринимать мои комментарии:</font><br>
<font color='DarkBlue'>Синий текст &mdash; просто текст комментария.</font><br>
<font color='DarkGreen'>👌 Зеленый текст &mdash; все отлично.</font><br>
<font color='DarkMagenta'>👉 Фиолетовый текст &mdash; сделано все правильно, однако есть рекомендации, на что стоит обратить внимание. Реализованные рекомендации позволят нам наработать опыт решения задачи разными способами или посмотреть на задачу под иным углом.</font><br>
<font color='Red'>🛠 Красный текст &mdash; есть недочеты, они иногда бывают. </font><br><font color='DarkBlue'>Любая ошибка это возможность посмотреть на задачу с другой стороны и освоить новые знания, по этому не надо расстраиваться, если они есть.</font><br>
<font color='DarkBlue'>Обращаю внимание, что комментарии ревьюера после проверки удалять нельзя они помогают отследить логику исправления или изменения проекта. Это дополнительный навык при обучении. Свой исправляемый код лучше тоже не удалять, а ремить #, это позволит акцентировать изменения в ходе подготовки окончательного варианта проекта.<br>
Ответы на мои комментарии лучше тоже помечать. 
Например: </font><font color='Blue'><b>Комментарий студента</b></font><br>
<font color='DarkBlue'><b>Теперь посмотрим, что получилось!</b></font>

# Проект спринта 6

- Автор: Андрей Творогов
- Дата: 12.01.2026

### Цели и задачи проекта

Цель проекта - подготовка и предварительный анализ данных по развитию индустрии видеоигр с 2000 по 2013 год.
Задачи, которые необходимо выполнить для достижения цели:
- изучение структуры данных, проверка наличия пропусков и дубликатов, соответствия типов данных типам столбцов в датасете;
- приведение названия столбцов к единому виду (snake case), типов столбцов к соответствию с типом их содержимого;
- оценка количества пропусков и дубликатов в данных;
- изучение пропусков и дубликатов, определение их типов и исправление данных там, где это возможно;
- отбор данных, относящихся к целевому периоду, из предоставленного датасета;
- категоризация оценок игр, выставленных пользователями и критиками.

### Описание данных

Данные `/datasets/new_games.csv` представлены в формате CSV (comma-separated values) и содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:

* `Name` — название игры.
* `Platform` — название платформы.
* `Year of Release` — год выпуска игры.
* `Genre` — жанр игры.
* `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
* `EU sales` — продажи в Европе (в миллионах проданных копий).
* `JP sales` — продажи в Японии (в миллионах проданных копий).
* `Other sales` — продажи в других странах (в миллионах проданных копий).
* `Critic Score` — оценка критиков (от 0 до 100).
* `User Score` — оценка пользователей (от 0 до 10).
* `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

1. <a href="#1">Загрузка данных и знакомство с ними</a>
2. <a href="#2">Проверка ошибок данных и их предобработка</a>
    - <a href="#2_1">2.1. Названия, или метки, столбцов датафрейма</a>
    - <a href="#2_2">2.2. Типы данных</a>
    - <a href="#2_3">2.3. Наличие пропусков в данных</a>
    - <a href="#2_4">2.4. Обработка пропусков в столбцах `user_score`, `critic_score`, `rating`</a>
    - <a href="#2_5">2.5. Явные и неявные дубликаты в данных</a>
    - <a href="#2_6">2.6. Выводы</a>
3. <a href="#3">Фильтрация данных</a>
3. <a href="#4">Категоризация данных</a>
3. <a href="#5">Итоговый вывод</a>

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Хорошо, сделано описание проекта. 👍 Оно помогает определиться над дальнейшей работой над проектом.</font>

---
## 1. Загрузка данных и знакомство с ними <a id="1"></a>

- Загрузка необходимых библиотек Python и данных датасета `/datasets/new_games.csv`:


In [1]:
import pandas as pd
import numpy as np

# Выгружаем данные в датафрейм
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkMagenta'>👉 Лучше разделять по разным ячейкам загрузку библиотек и датасета.</font><br>
<font color='DarkGreen'>👌 Очень хорошо! Загрузка датасета непосредственно с сервера и все равно от куда запускается тетрадка. 👍</font>

- Пример данных (первые 5 строк датасета):

In [2]:
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


- Сведения о датасете (результат вызова метода `info()`):

In [3]:
total_size = len(df)
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  str    
 1   Platform         16956 non-null  str    
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  str    
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  str    
 6   JP sales         16956 non-null  str    
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  str    
 10  Rating           10085 non-null  str    
dtypes: float64(4), str(7)
memory usage: 1.4 MB


- Количество пропусков по столбцам в абсолютных единицах и их доля в процентах:

In [4]:
pd.DataFrame({'Count': df.isna().sum(), 'Percent': round(df.isna().mean()*100, 2)})

,Count,Percent
Name,2,0.01
Platform,0,0.00
Year of Release,275,1.62
Genre,2,0.01
NA sales,0,0.00
EU sales,0,0.00
JP sales,0,0.00
Other sales,0,0.00
Critic Score,8714,51.39
User Score,6804,40.13


**Промежуточные выводы:**
- В предоставленных данных 16956 записей, данные в столбцах при поверхностном анализе соответствуют их наименованию;
- типы данных определяются недостаточно точно: все столбцы имеют тип `object` или `float64`, однако столбцы `Name`, `Platform`, `Genre`, `Rating` хранят строковые данные, столбец `Year of Release` - целочисленные, `EU sales`, `JP sales` - вещественные;
- количество пропусков в столбцах `Name` и `Year of Release` невелико (2 и 275, что составляет 0.01% и 1.62% соответственно), однако в столбцах `Critic Score`, `User Score`, `Rating` их 8714, 6804 и 6871, что составляет 51.39%, 40.13%, 40.52% соответственно. Эти пропуски должны быть проанализированы и, по возможности, заполнены в соответствии с выбранной стратегией.
- Названия столбцов содержат заглавные буквы и пробелы, что не вполне удобно для обработки и не соответствует соглашениям об именах в Python. Предлагается привести названия столбцов к snake case.

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Хорошо, сделали первичный анализ данных перед предобработкой.</font>

---
## 2. Проверка ошибок в данных и их предобработка <a id="2"></a>

### 2.1. Названия, или метки, столбцов датафрейма <a id="2_1"></a>

Названия всех столбцов датафрейма:

In [5]:
df.columns.to_list()

['Name',
 'Platform',
 'Year of Release',
 'Genre',
 'NA sales',
 'EU sales',
 'JP sales',
 'Other sales',
 'Critic Score',
 'User Score',
 'Rating']

Приведем названия всех столбцов к snake case:

In [6]:
df.columns = df.columns.str.lower().str.replace(' ', '_') # df.columns = [col.lower().replace(' ', '_') for col in df.columns]
df.columns.to_list()

['name',
 'platform',
 'year_of_release',
 'genre',
 'na_sales',
 'eu_sales',
 'jp_sales',
 'other_sales',
 'critic_score',
 'user_score',
 'rating']

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Хорошо! Приводим названия колонок к общепринятому стилю. 😉 <br>
А можно просто методом вот так:</font><br>
`df.columns = df.columns.str.lower().str.replace(' ', '_')`

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkMagenta'>👉 Лучше перед обработкой данных делать контроль размера датасета до и после обработки. Примерно вот так:</font>

In [7]:
# Комментарий ревьюера
# создаем копию датасета до преобразования для возможности проверить сделанные изменения после предобработки
tmp = df.copy() 
len(tmp)

16956

### 2.2. Типы данных <a id="2_2"></a>

- Столбцы со строковыми данными импортировались с типом `object` (обычное поведение библиотеки `pandas` по умолчанию);
- Год выпуска в исходных данных имеет дробную часть, поэтому столбец `Year of Release` при импорте получил тип `float64`;

Необходимо выяснить причины, по которым столбцы `eu_sales`, `jp_sales`, `user_score` получили тип `object`:

In [8]:
# Пробуем преобразовать столбцы к числовому типу и смотрим, почему это не получается
print('Столбец eu_sales:')
try:
    pd.to_numeric(df['eu_sales'])
except ValueError as err:
    print('Ошибка преобразования типа:', err)

print('Столбец jp_sales:')
try:
    pd.to_numeric(df['jp_sales'])
except ValueError as err:
    print('Ошибка преобразования типа:', err)

print('Столбец user_score:')
try:
    pd.to_numeric(df['user_score'])
except ValueError as err:
    print('Ошибка преобразования типа:', err)

Столбец eu_sales:
Ошибка преобразования типа: Unable to parse string "unknown" at position 446
Столбец jp_sales:
Ошибка преобразования типа: Unable to parse string "unknown" at position 467
Столбец user_score:
Ошибка преобразования типа: Unable to parse string "tbd" at position 119


<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Правильно, что проверили ошибки значений в столбцах с типом `object`, которые по смыслу должны быть числовыми.</font>

Столбцы `eu_sales`, `jp_sales`, `user_score` содержат текстовые данные: 'unknown' и 'tbd' (вероятно, to be defined). Эти значения могут быть преобразованы в `NaN` без потери значимой информации.

Преобразуем типы столбцов следующим образом:
- `name`: `str`
- `platform`: `str`
- `year_of_release`: `int16`
- `genre`: `str`
- `eu_sales`: `float64`
- `jp_sales`: `float64`
- `user_score`: `float64`
- `rating`: `str`

**Примечание:** Столбец `year_of_release` содержит пропуски, из-за чего его нельзя преобразовать к целочисленному типу. Поэтому удалим соответствующие строки (1.6% от датасета).

In [9]:
# def calc_year(row):
#     year = row['year_of_release']
#     if pd.notna(year):
#         return year
#     return df[df['platform'] == row['platform']]['year_of_release'].median()

# # Преобразование типа столбца к nullable float64               
# df['year_of_release'] = df['year_of_release'].astype('float64')
# # Заполнение пропусков
# df['year_of_release'] = df.apply(calc_year, axis=1)

# Удаляем пропуски в year_of_release
df = df.dropna(subset=['year_of_release'])

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkMagenta'>👉 Обработка пропусков должна проводиться после их анализа в следующем разделе проекта.</font><br>
<font color='DarkMagenta'>👉 Зачем получать "медианную температуру по больнице" по пропускам года релиза, более того некоторые игры выходили на одной платформе под одним названием в разные года. Пропусков всего 1,6%, удалить их и забыть.</font>

<font color='Blue'><b>Комментарий студента</b></font><br>
<font color='Blue'>Сначала сделал с удалением, потом подумал, что, поскольку срок жизни платформы ограничен, можно заполнить. Теперь вернул первоначальный вариант. Приходится удалять данные в этом разделе, потому что без этого не получится преобразовать тип к int.</font>

In [10]:
# Преобразование типов столцов
df = df.astype({'name': 'string',
                'platform': 'string',
                'year_of_release': 'int',
                'genre': 'string',
                'rating': 'string'})

# Отдельно преобразуем столбцы с данными смешанных типов
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors = 'coerce')
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors = 'coerce')
df['user_score'] = pd.to_numeric(df['user_score'], errors = 'coerce')

df.info()

<class 'pandas.DataFrame'>
Index: 16681 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  string 
 1   platform         16681 non-null  string 
 2   year_of_release  16681 non-null  int64  
 3   genre            16679 non-null  string 
 4   na_sales         16681 non-null  float64
 5   eu_sales         16675 non-null  float64
 6   jp_sales         16677 non-null  float64
 7   other_sales      16681 non-null  float64
 8   critic_score     8085 non-null   float64
 9   user_score       7558 non-null   float64
 10  rating           9901 non-null   string 
dtypes: float64(6), int64(1), string(4)
memory usage: 1.5 MB


<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Хорошо, преобразовали типы данных и сделали проверку преобразований.</font>

### 2.3. Наличие пропусков в данных <a id="2_3"></a>

Еще раз посчитаем количество пропусков в каждом столбце в абсолютных и относительных значениях:


In [11]:
pd.DataFrame({'Count': df.isna().sum(), 'Percent': round(df.isna().mean()*100, 2)})

,Count,Percent
name,2,0.01
platform,0,0.00
year_of_release,0,0.00
genre,2,0.01
na_sales,0,0.00
eu_sales,6,0.04
jp_sales,4,0.02
other_sales,0,0.00
critic_score,8596,51.53
user_score,9123,54.69


<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkMagenta'>👉 Можно вот так вывести информацию о пропусках:</font>

<font color='Blue'><b>Комментарий студента</b></font><br>
<font color='Blue'>Благодарю!</font>

In [12]:
# Комментарий ревьюера
def show_missing_stats(tmp0):
    """
    Функция для отображения статистики пропущенных значений в DataFrame.
    """
    missing_stats = pd.DataFrame({
        'Кол-во пропусков': tmp0.isnull().sum(),
        'Доля пропусков': tmp0.isnull().mean()
    })
    missing_stats = missing_stats[missing_stats['Кол-во пропусков'] > 0]
    
    if missing_stats.empty:
        return "Пропусков в данных нет"
    
    # Форматируем при выводе через Styler
    return (missing_stats.style.format({'Доля пропусков': '{:.4f}'}).background_gradient(cmap='coolwarm'))
show_missing_stats(df)

,Кол-во пропусков,Доля пропусков
name,2,0.0001
genre,2,0.0001
eu_sales,6,0.0004
jp_sales,4,0.0002
critic_score,8596,0.5153
user_score,9123,0.5469
rating,6780,0.4065


<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='Red'>🛠 Вот, что написано в задании: <br><b>Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.</b> <br>Предположений о причинах пропусков и предложений по их обработке не сделано.</font>

<font color='Blue'><b>Комментарий студента</b></font><br>
<font color='Blue'>Предположения относительно пропусков в столбцах с оценками - в следующем разделе. Дополнил выводы предположениями о причинах появления пропусков.</font>

- Пропуски в столбцах `eu_sales` и `jp_sales` могут быть случайными, заполняем их средними значениями с учетом платформы и года выпуска:

In [13]:
def approx_sales(row):
    eu_sales = row['eu_sales']
    jp_sales = row['jp_sales']
    
    if pd.isna(eu_sales) or pd.isna(jp_sales):
        group = df[(df['platform'] == row['platform']) & 
                   (df['year_of_release'] == row['year_of_release'])]
        
    if pd.isna(eu_sales):
        eu_sales = group['eu_sales'].mean()
    if pd.isna(jp_sales):
        jp_sales = group['jp_sales'].mean()

    return eu_sales, jp_sales

rv = df.apply(approx_sales, axis=1)
df[['eu_sales', 'jp_sales']] = pd.DataFrame(rv.tolist(), index=rv.index)

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Хорошо, выполнено задание по замене пропусков средним по столбцам платформ и году релиза.</font>

### 2.4. Обработка пропусков в столбцах `user_score`, `critic_score`, `rating` <a id="2_4"></a>
Попробуем выяснить зависимости пустых значений от значений в других столбцах (определение типа пропусков - MCAR или MNAR)

In [14]:
# Добавление вспомогательных столбцов
df['summary_sales'] = df['na_sales'] + df['eu_sales'] + df['jp_sales'] + df['other_sales'] # Суммарные продажи
df['critic_score_missed'] = df['critic_score'].isna()       # Флаг отсутствия значения в столбце `critic_score`
df['user_score_missed'] = df['user_score'].isna()           # Флаг отсутствия значения в столбце `user_score`
df['jp_sales_category'] = pd.cut(df['jp_sales'], bins=20)   # Категоризация значений в столбце `jp_sales`
# Категоризация по значениям в столбце `summary_sales`
# Квантильная разбивка - в каждой группе примерно поровну строк
df['sum_sales_category'] = pd.qcut(df['summary_sales'], q=int(10), duplicates='drop')

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkMagenta'>👉 Не понятные манипуляции с данными:
- Если создавать столбец с суммой продаж, то почему не используются продажи в столбце `other_sales`?
- Зачем создаются столбцы с признаками пропусков в столбцах отзывов? Если хотим делать выборки для анализа без пропусков в отзывах, то это можно сделать и без лишних столбцов.
- С какой целью категоризируются данные продаж в в Японии и суммарные продажи?</font>

<font color='Blue'><b>Комментарий студента</b></font><br>
- <font color='Blue'>Cтолбец `other_sales` пропустил, сейчас добавил;</font>
- <font color='Blue'>столбцы с признаками добавил для того, чтобы было понятно, какие значения были заполнены искусственными данными (сейчас уже неважно). Но они используются далее в `crosstab`;</font>
- <font color='Blue'>объемы продаж категоризировал для того, чтобы была видна зависимость между продажами и количеством пропусков. Она и видна в таблицах. Как еще ее обнаружить?</font>

- Доля пропусков в зависимости от платформы:

In [15]:
# Доля пропусков в зависимости от платформы
pd.crosstab(df['platform'], [df['critic_score_missed'], df['user_score_missed']], normalize='index')

critic_score_missed     False               True           
user_score_missed       False     True      False     True 
platform                                                   
2600                 0.000000  0.000000  0.000000  1.000000
3DO                  0.000000  0.000000  0.000000  1.000000
3DS                  0.300766  0.019157  0.028736  0.651341
DC                   0.269231  0.000000  0.000000  0.730769
DS                   0.218910  0.114113  0.022357  0.644620
GB                   0.000000  0.000000  0.000000  1.000000
GBA                  0.290557  0.236077  0.016949  0.456416
GC                   0.641166  0.163934  0.016393  0.178506
GEN                  0.000000  0.000000  0.000000  1.000000
GG                   0.000000  0.000000  0.000000  1.000000
N64                  0.000000  0.000000  0.000000  1.000000
NES                  0.000000  0.000000  0.000000  1.000000
NG                   0.000000  0.000000  0.000000  1.000000
PC                   0.718107  0.011317  0.069959  0.200617
PCFX                 0.000000  0.000000  0.000000  1.000000
PS                   0.127483  0.036424  0.004967  0.831126
PS2                  0.535747  0.063138  0.038533  0.362581
PS3                  0.590226  0.023308  0.066165  0.320301
PS4                  0.630380  0.007595  0.020253  0.341772
PSP                  0.326733  0.055281  0.023927  0.594059
PSV                  0.278802  0.002304  0.055300  0.663594
SAT                  0.000000  0.000000  0.000000  1.000000
SCD                  0.000000  0.000000  0.000000  1.000000
SNES                 0.000000  0.000000  0.000000  1.000000
TG16                 0.000000  0.000000  0.000000  1.000000
WS                   0.000000  0.000000  0.000000  1.000000
Wii                  0.371648  0.068199  0.052874  0.507280
WiiU                 0.605442  0.006803  0.074830  0.312925
X360                 0.698159  0.028823  0.068054  0.204964
XB                   0.704156  0.174817  0.006112  0.114914
XOne                 0.665339  0.015936  0.067729  0.250996

- Доля пропусков в зависимости от года выпуска:

In [16]:
# Доля пропусков в зависимости от года выпуска
pd.crosstab(df['year_of_release'], [df['critic_score_missed'], df['user_score_missed']], normalize='index')

critic_score_missed     False               True           
user_score_missed       False     True      False     True 
year_of_release                                            
1980                 0.000000  0.000000  0.000000  1.000000
1981                 0.000000  0.000000  0.000000  1.000000
1982                 0.000000  0.000000  0.000000  1.000000
1983                 0.000000  0.000000  0.000000  1.000000
1984                 0.000000  0.000000  0.000000  1.000000
1985                 0.071429  0.000000  0.000000  0.928571
1986                 0.000000  0.000000  0.000000  1.000000
1987                 0.000000  0.000000  0.000000  1.000000
1988                 0.066667  0.000000  0.000000  0.933333
1989                 0.000000  0.000000  0.000000  1.000000
1990                 0.000000  0.000000  0.000000  1.000000
1991                 0.000000  0.000000  0.000000  1.000000
1992                 0.023256  0.000000  0.000000  0.976744
1993                 0.000000  0.000000  0.000000  1.000000
1994                 0.008264  0.000000  0.000000  0.991736
1995                 0.000000  0.000000  0.000000  1.000000
1996                 0.029963  0.000000  0.000000  0.970037
1997                 0.047782  0.010239  0.013652  0.928328
1998                 0.067708  0.005208  0.013021  0.914062
1999                 0.090909  0.026393  0.002933  0.879765
2000                 0.285714  0.117647  0.014006  0.582633
2001                 0.533605  0.144603  0.002037  0.319756
2002                 0.548272  0.207390  0.010727  0.233611
2003                 0.643853  0.110266  0.021546  0.224335
2004                 0.623865  0.111543  0.015564  0.249027
2005                 0.598947  0.097895  0.022105  0.281053
2006                 0.521569  0.092157  0.023529  0.362745
2007                 0.493432  0.087028  0.032020  0.387521
2008                 0.417993  0.083737  0.045675  0.452595
2009                 0.386207  0.068276  0.044828  0.500690
2010                 0.343237  0.054730  0.029711  0.572322
2011                 0.408181  0.029591  0.030461  0.531767
2012                 0.485075  0.000000  0.028358  0.486567
2013                 0.501812  0.001812  0.057971  0.438406
2014                 0.440068  0.008562  0.125000  0.426370
2015                 0.364379  0.006536  0.125817  0.503268
2016                 0.447732  0.009862  0.071006  0.471400

- Доля пропусков в зависимости от категории продаж:

In [17]:
df['sum_sales_category'] = df['sum_sales_category'].astype(str) # Сбрасываем CategoricalIndex
pd.crosstab(df['sum_sales_category'], [df['critic_score_missed'], df['user_score_missed']], normalize='index')

critic_score_missed     False               True           
user_score_missed       False     True      False     True 
sum_sales_category                                         
(-0.001, 0.02]       0.205802  0.069513  0.037767  0.686918
(0.02, 0.05]         0.243128  0.103318  0.027014  0.626540
(0.05, 0.08]         0.322798  0.086752  0.030935  0.559516
(0.08, 0.11]         0.353630  0.104606  0.026542  0.515222
(0.11, 0.17]         0.393902  0.089634  0.034756  0.481707
(0.17, 0.25]         0.447964  0.069570  0.035633  0.446833
(0.25, 0.38]         0.470149  0.056592  0.046020  0.427239
(0.38, 0.6]          0.534483  0.038793  0.049261  0.377463
(0.6, 1.2]           0.609581  0.035329  0.032335  0.322754
(1.2, 82.54]         0.646459  0.009004  0.027011  0.317527

- Культура оценивания игр в Японии отличается от европейской и американской, поэтому отдельно оценим количество пропусков в зависимости от продаж в Японии:

In [18]:
# Количество пропусков в зависимости от продаж в Японии
pd.crosstab(df['jp_sales_category'], [df['critic_score_missed'], df['user_score_missed']])

critic_score_missed False       True       
user_score_missed   False True  False True 
jp_sales_category                          
(-0.0102, 0.511]     6783  1104   578  7660
(0.511, 1.022]        109     2     1   208
(1.022, 1.533]         47     0     0    73
(1.533, 2.044]          9     0     0    26
(2.044, 2.555]          8     0     0    18
(2.555, 3.066]          7     0     0     8
(3.066, 3.577]          2     0     0     8
(3.577, 4.088]          5     0     0     5
(4.088, 4.599]          5     0     0     4
(4.599, 5.11]           1     0     0     1
(5.11, 5.621]           2     0     0     1
(5.621, 6.132]          0     0     0     2
(6.132, 6.643]          1     0     0     0
(6.643, 7.154]          0     0     0     1
(7.154, 7.665]          0     0     0     1
(9.709, 10.22]          0     0     0     1

### Выводы:
- Наибольшее количество пропусков содержится в столбцах `user_score`, `critic_score`, `rating`- 51.53%, 54.69%, 40.65% соответственно;
- Пропуски относятся к типу MNAR и наиболее характерны для следующих игр:
    - игр для платформ 3DS, DC, PS, PSP, PSV (портативных или выпущенных до 2000 года)
    - вообще для игр выпущенных до 2000 года (почти 100% отсутствие оценок);
    - игр с низким объемом продаж;
    - игр, продаваемых в Японии.

Появление пропусков можно объяснить так:
- Онлайн-платформы с оценками игр начали получать популярность в начале 2000-х, поэтому большинство игр, выпущенных до 2000 г., не имеют рейтингов.
- Портативные платформы хуже освещаются на игровых онлайн-порталах, имеют свою закрытую экосистему, в которой долгое время не было пользовательских оценок; также они более распространены в Японии, где культура восприятия игр и оценивания отличается от американской и европейской.
- Игры с низким объемом продаж (малобюджетные или нишевые) могут не получать рецензий со стороны профессиональных критиков. Как следствие, они менее популярны у пользователей, разброс оценок может быть большим, а их количество - недостаточным для формирования итоговой оценки.

Оценка не является следствием случайных наблюдений, а зависит от многих факторов, которые не могут быть формализованы и не сводятся к имеющейся в датасете информации, поэтому правильно будет заполнить пропуски в столбцах `user_score` и `critic_score` индикатором -1:

In [19]:
df.loc[df['user_score'].isna(), 'user_score'] = -1
df.loc[df['critic_score'].isna(), 'critic_score'] = -1

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='Red'>🛠 Почему считаем, что такая замена пропусков колонок отзывов корректна?</font><br>
<font color='DarkMagenta'>👉 Проще поставить на них заглушку (значение-индикатор) в виде отрицательной цифры.</font><br>
<font color='DarkMagenta'>👉 Ошибка в подходе к заполнению пропусков в колонках отзывов средним/медианой/модой или синтетическими данными, извлекаемыми из данных других колонок, связана с тем, что данные этих колонок не являются набором случайных наблюдений. <br>Если бы здесь была выборка оценок пользователей по отзывам на одну игру, то ее можно было считать случайными данными. Один пользователь поставил отметку 1, другой 10. Одному игра не понравилась, другому понравилсь. Пропуски в такой выборке можем заменить средним/медианой/модой, так как все данные относятся к одной и той же игре. <br>В данном случае отзыв как раз и является таким усреднением выборки оценок пользователей по одной игре и является признаком именно этой игры, не зависящей от жанра, платформы и т.д. Более того, если внимательно посмотреть на датасет, то можем найти примеры, когда одна игра на разных платформах имеет двукратное различие в оценках пользователей. Возможно, что не на всех платформах можно достичь высокого качества графики, скорости и т.д.<br>Так же формируется и отзыв критиков. Определенное количество критиков протестировало игру, выставило свою оценку, которую потом усреднили и присвоили конкретной игре. То есть отзыв критиков стал интегрированным признаком конкретной игры и не зависит от жанра, платформы и т.д.<br>
По этому пропуски отзывов нельзя заполнять синтетическими данными извлекаемыми из оценок других игр.<br>
Задача при анализе датасета и обработке пропусков заключается определить, что собой представляют данные и какие реальные бизнес-процессы они описывают. И в зависимости от этого принимать решение об обработке.</font>

<font color='Blue'><b>Комментарий студента</b></font><br>
<font color='Blue'>Исправил, благодарю за разъяснение</font>

- Пропуски в столбцах `rating` заполняем индикатором `Undefined`;
- пропуски в столбцах `name` и `genre` удалим

In [20]:
df.loc[df['rating'].isna(), 'rating'] = 'Undefined'
df = df.dropna(subset=['name', 'genre'])

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkMagenta'>👉 Не совсем обоснованное решение по пропускам колонки имен игр и жанров. В данных всего два пропуска, да еще в этих строках пропуски по отзывам и рейтингу. То есть именно по тем данным, которые будут анализироваться в проекте. Получается, что поставив заглушку (значение-индикатор) на эти пропуски, тем самым не сохраним данные строчки для анализа, а внесем в них больше неопределенности. Может проще удалить?</font><br>
<font color='DarkGreen'>👌 Правильно поставлена заглушка (значение-индикатор) на пропуски в колонке рейтинга. Если посмотреть в интернете, что такое рейтинг ESRB, то узнаем, что это рейтинг для США и Канады. А значит пропуск не связан с ошибкой заполнения данных, просто не только для Северной Америки выпускаются игры.</font>

<font color='Blue'><b>Комментарий студента</b></font><br>
<font color='Blue'>Исправил, благодарю за разъяснение</font>

In [21]:
# Комментарий ревьюера
# Посмотрим какие пропуски остались
show_missing_stats(df)

'Пропусков в данных нет'

### 2.5. Явные и неявные дубликаты в данных <a id="2_5"></a>
Необходимо выявить наличие явных и неявных дубликатов в данных и ликвидировать их. В первую очередь исследуем значения в столбцах, содержащих категориальные данные.

- Проверка категориальных значений в столбце `genre`:

In [22]:
df['genre'].unique()

<StringArray>
[      'Sports',     'Platform',       'Racing', 'Role-Playing',
       'Puzzle',         'Misc',      'Shooter',   'Simulation',
       'Action',     'Fighting',    'Adventure',     'Strategy',
         'MISC', 'ROLE-PLAYING',       'RACING',       'ACTION',
      'SHOOTER',     'FIGHTING',       'SPORTS',     'PLATFORM',
    'ADVENTURE',   'SIMULATION',       'PUZZLE',     'STRATEGY']
Length: 24, dtype: string

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Хорошо, сделали проверку уникальных значений в столбце с категориальными данными.</font>

Встречаются неявные дубликаты, отличающиеся регистром символов. Приведем все названия жанров к нижнем регистру:

In [23]:
df['genre'] = df['genre'].str.lower()

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Правильно сделано преобразование регистров данных, тем самым удалены неявные дубликаты.</font>

- Проверка категориальных значений в столбце `platform`:

In [24]:
df['platform'].unique()

<StringArray>
[ 'Wii',  'NES',   'GB',   'DS', 'X360',  'PS3',  'PS2', 'SNES',  'GBA',
  'PS4',  '3DS',  'N64',   'PS',   'XB',   'PC', '2600',  'PSP', 'XOne',
 'WiiU',   'GC',  'GEN',   'DC',  'PSV',  'SAT',  'SCD',   'WS',   'NG',
 'TG16',  '3DO',   'GG', 'PCFX']
Length: 31, dtype: string

Все данные корректны, дубликаты отсутствуют.

- Проверка категориальных данных в столбце `rating`:

In [25]:
df['rating'].unique()

<StringArray>
['E', 'Undefined', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP']
Length: 9, dtype: string

Рейтинг `K-A` является устаревшим: они использовался с 1994 по 1998 г., затем был заменен на `E`. Заменим его в данных на `E`, хотя это необязательно, так как нас будут интересовать только игры, выпущенные после 2000 г.

In [26]:
df.loc[df['rating'] == 'K-A', 'rating'] = 'E'

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Молодец! Сделали замену устаревшего рейтинга. 👍</font>

- Проверка категориальных данных в столбце `year_of_release`:

In [27]:
years = df['year_of_release'].unique()
years.sort()
years

array([1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990,
       1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001,
       2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012,
       2013, 2014, 2015, 2016])

Данные корректны, дубликаты отсутствуют.

Помимо неявных, данные могли содержать явные дубликаты. Их количество могло стать еще большим после унификации категориальных значений в столбце `genre`.
Посчитаем явные дубликаты по всем столбцам и удалим их:

In [28]:
print('Количество явных дубликатов:', df.duplicated().sum())
df = df.drop_duplicates()

Количество явных дубликатов: 235


<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Хорошо, что удалены явные дубликаты, в дальнейшем они искажают результаты анализа.</font>

Сравним количество строк в исходном датасете с их количеством после предварительной обработки:

In [29]:
print(f'Первоначальное количество строк: {total_size}')
print(f'Количество строк после предварительной обработки: {len(df)}')
score_misses = ((df['critic_score'] == -1) | (df['user_score'] == -1)).sum()
print(f'Количество строк с индикаторами -1 в столбцах user_score или critic_score: {score_misses} ({round(score_misses / len(df) * 100.0, 2)}%)')
print(f'Удалено строк: {total_size - len(df)} ({round((total_size-len(df)) / total_size * 100.0, 2)}%)')

Первоначальное количество строк: 16956
Количество строк после предварительной обработки: 16444
Количество строк с индикаторами -1 в столбцах user_score или critic_score: 9550 (58.08%)
Удалено строк: 512 (3.02%)


<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Хорошо, сделали проверку удаленных данных.</font>

In [30]:
# Комментарий ревьюера
# Проверим сколько удалено строк датасета
a, b = len(tmp), len(df)
print(" Было строк в исходном датасете", a,
      '\n', "Осталось строк в датасете после обработки", b,
      '\n', "Удалено строк в датасете после обработки", a-b,
      '\n', "Процент потерь", round((a-b)/a*100, 2))

 Было строк в исходном датасете 16956 
 Осталось строк в датасете после обработки 16444 
 Удалено строк в датасете после обработки 512 
 Процент потерь 3.02


### 2.6. Выводы <a id="2_6"></a>
- Для анализа был представлен датасет в формате CSV объемом 16956 строк.
- При импорте датасета типы некоторых столбцов были определены неверно. Их типы были явно приведены к типу содержащихся в них данных;
- В исходный датасет были добавлены следующие столбцы:
    1. `summary_sales` - суммарные продажи игры во всех регионах;
    2. `critic_score_missed` - флаг пропуска значения в столбце `critic_score`;
    3. `user_score_missed` - флаг пропуска значения в столбце `user_score`;
    4. `jp_sales_category` - категория объема продаж в Японии (всего 20 категорий);
    5. `sum_sales_category` - категория суммарного объема продаж (была произведена квантильная разбивка всего датасета по суммарному объему продаж на примерно равные по количеству строк категории);
- В исходных данных содержалось значительное количество (около 50%) пропусков в значимых для анализа столбцах `user_score` и `critic_score`. Пропуски в этих полях были заменены на индикатор **-1**;
- Пропуски в строковых полях `name`, `genre` составляют менее 1%; соответствующие строки были удалены;
- Пропуски в столбце `rating` также были заменены на индикатор `Undefined`;
- Строки с пропущенным годом выпуска (275 или 1,6%) были удалены;
- При анализе неявных дубликатов было обнаружено, что значения в столбце `genre` дублируются с точностью до регистра символов. Все значения были приведены к нижнему регистру;
- Был выполнен анализ датасета на наличие явных дубликатов с их удалением.<br><br>
В результате выполненных действий были удалены 512 строк (3.02% от исходного датасета). Количество строк в полученном датасете составляет 16444.

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Хорошо, сделали вывод по предобработке данных.</font>

## 3. Фильтрация данных <a id="3"></a>

Для анализа требуются данные с 2000 по 2013 год. Отфильруем данные по этим условиям и сохраним их в новый датасет `df_actual`:

In [31]:
df_actual = pd.DataFrame(df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)])
print('Количество записей за 2000-2013 гг.:', len(df_actual))
df_actual.info()

Количество записей за 2000-2013 гг.: 12781
<class 'pandas.DataFrame'>
Index: 12781 entries, 0 to 16954
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   name                 12781 non-null  string  
 1   platform             12781 non-null  string  
 2   year_of_release      12781 non-null  int64   
 3   genre                12781 non-null  string  
 4   na_sales             12781 non-null  float64 
 5   eu_sales             12781 non-null  float64 
 6   jp_sales             12781 non-null  float64 
 7   other_sales          12781 non-null  float64 
 8   critic_score         12781 non-null  float64 
 9   user_score           12781 non-null  float64 
 10  rating               12781 non-null  string  
 11  summary_sales        12781 non-null  float64 
 12  critic_score_missed  12781 non-null  bool    
 13  user_score_missed    12781 non-null  bool    
 14  jp_sales_category    12781 non-null  catego

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Правильно отфильтровали данные по годам релиза. А проверку результата фильтрации можно провести еще вот так:</font>

In [32]:
# Комментарий ревьюера
df_actual.year_of_release.sort_values().unique()

array([2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
       2011, 2012, 2013])

## 4. Категоризация данных <a id="4"></a>
    
Проведем категоризацию данных:
- Разделение всех игр по оценкам пользователей и выделение таких категорий: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [33]:
df_actual['user_score_category'] = pd.cut(
    df_actual['user_score'],
    bins=[-1, 0, 3, 8, 10.01],
    labels=['Undefined', 'Низкая', 'Средняя', 'Высокая'],
    right=False
)

- Разделение всех игр по оценкам критиков и выделение таких категорий: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [34]:
df_actual['critic_score_category'] = pd.cut(
    df_actual['critic_score'],
    bins=[-1, 0, 30, 80, 100.01],
    labels=['Undefined', 'Низкая', 'Средняя', 'Высокая'],
    right=False
)

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Молодец, правильно сделана категоризация данных отзывов.<br> Хорошо, что оценки отзывов, на которые поставлены заглушки, вывели в отдельную категорию.</font>

- Группировка данных по выделенным категориям и подсчет количества игр в каждой категории:

In [35]:
df_actual[['user_score_category', 'name']].groupby('user_score_category', observed=False).agg('count')

,name
user_score_category,
Undefined,6298
Низкая,116
Средняя,4081
Высокая,2286


In [36]:
df_actual[['critic_score_category', 'name']].groupby('critic_score_category', observed=False).agg('count')

,name
critic_score_category,
Undefined,5612
Низкая,55
Средняя,5422
Высокая,1692


<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='Red'>🛠 Вот, все пропуски, которые были заменены средними попали в категорию "средняя оценка", что делает дальнейшиый анализ не валидным.</font>

<font color='Blue'><b>Комментарий студента</b></font><br>
<font color='Blue'>Исправил.</font>

- Выделение топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [37]:
df_actual[['platform', 'name']].groupby('platform', observed=False).agg('count').sort_values(by='name', ascending=False).head(7)

,name
platform,
PS2,2127
DS,2120
Wii,1275
PSP,1180
X360,1121
PS3,1087
GBA,811


<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Правильно выведен ТОП-7 платформ по количеству выпущенных игр.</font>

## 5. Итоговый вывод <a id="5"></a>
В процессе работы над проектом была выполнена предварительная обработка данных: переименование и уточнение типов данных столбцов, определение и заполнение пропусков, удаление дубликатов. Наибольшее количество пропусков было обнаружено в важных для анализа столбцах `user_score`, `critic_score`, а также в столбце `rating`. Пропущенные оценки игр были заполнены индикаторами **-1**.<br><br>
Для дальнейшей обработки был сформирован срез данных `df_actual`, включающий игры, выпущенные с 2000 по 2013 год.
В процессе категоризации среза данных к нему были добавлены поля, содержащие категорию оценок (`Низкая`, `Средняя`, `Высокая` и `Undefined` для пропущенных значений).
<br>
Ниже перечислены все добавленные относительно исходного датасета столбцы:
1. `summary_sales` - суммарные продажи игры во всех регионах;
2. `critic_score_missed` - флаг пропуска значения в столбце `critic_score`;
3. `user_score_missed` - флаг пропуска значения в столбце `user_score`;
4. `jp_sales_category` - категория объема продаж в Японии (всего 20 категорий);
5. `sum_sales_category` - категория суммарного объема продаж;
6. `user_score_category` - категория оценки пользователями;
7. `critic_score_category` - категория оценки критиками.
<br><br>

Также был выполнен подсчет игр в каждой выделенной категории и определены 7 наиболее популярных игровых платформ.

<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Итоговый вывод написан в полном объеме.</font>
<br><font color='DarkMagenta'>👉 После исправления ошибок вывод необходимо будет откорректировать.</font>

<font color='DarkBlue'><b>Заключительный комментарий ревьюера</b></font><br><br>
<font color='DarkGreen'>👌 Андрей, проект отлично структурирован и проработан.<br>Так держать и дальше!<br></font>
<font color='Red'>🛠 Что нужно, что бы сделать проект лучше:</font><br>
<font color='Red'>🛠 Провести анализ пропусков и их причин, сделать промежуточные выводы. Переделать обработку пропусков в отзывах пользователей и критиков.<br>Обращаю внимание, что после исправления ошибок необходимо будет откорректировать выводы анализа и предобработки.</font><br>
<font color='DarkMagenta'>👉 При исправлении ошибок и недочетов будут приобретены дополнительный опыт и навыки!</font><br>
<font color='DarkGreen'>👌 Жду окончательный вариант! Успехов.</font>

<font color='Blue'><b>Комментарий студента</b></font><br>
<font color='Blue'>Исправил, благодарю за замечания и пояснения. Они были очень полезны.</font>